python/cuml/internals/api_decorators.py

```diff
@@ -39,646 +36,206 @@
-            available or called with the wrong number of arguments
-        """
-        self_val = None
-        input_val = None
-        target_val = None
-
-        if (self.has_self):
-            self_val = args[0]
-
-        if (self.needs_input):
-            # Check if its set to a string
-            if (isinstance(self.input_arg_to_use, str)):
-                input_val = kwargs[self.input_arg_to_use]
-
-            # If all arguments are set by name, then this can happen
-            elif (self.input_arg_to_use >= len(args)):
-                # Check for the name in kwargs
-                if (self.input_arg_to_use_name in kwargs):
-                    input_val = kwargs[self.input_arg_to_use_name]
-                else:
-                    raise IndexError(
-                        ("Specified input_arg idx: {}, and argument name: {}, "
-                         "were not found in args or kwargs").format(
-                             self.input_arg_to_use,
-                             self.input_arg_to_use_name))
-            else:
-                # Otherwise return the index
-                input_val = args[self.input_arg_to_use]
-
-        if (self.needs_target):
-            # Check if its set to a string
-            if (isinstance(self.target_arg_to_use, str)):
-                target_val = kwargs[self.target_arg_to_use]
-
-            # If all arguments are set by name, then this can happen
-            elif (self.target_arg_to_use >= len(args)):
-                # Check for the name in kwargs
-                if (self.target_arg_to_use_name in kwargs):
-                    target_val = kwargs[self.target_arg_to_use_name]
-                else:
-                    raise IndexError((
-                        "Specified target_arg idx: {}, and argument name: {}, "
-                        "were not found in args or kwargs").format(
-                            self.target_arg_to_use,
-                            self.target_arg_to_use_name))
+def _get_value(args, kwargs, name, index):
+    """Determine value for a given set of args, kwargs, name and index."""
+    try:
+        return kwargs[name]
+    except KeyError:
+        try:
+            return args[index]
+        except IndexError:
+            raise IndexError(
+                f"Specified arg idx: {index}, and argument name: {name}, "
+                "were not found in args or kwargs.")
+
+
+def _make_decorator_factory(
+    context_manager_cls: InternalAPIContextBase,
+    process_return=True,
+    needs_self: bool = False,
+    ** defaults,
+) -> typing.Callable[..., _DecoratorType]:
+
+    def factory(
+        input_arg: str = ...,
+        target_arg: str = ...,
+        get_output_type: bool = False,
+        set_output_type: bool = False,
+        get_output_dtype: bool = False,
+        set_output_dtype: bool = False,
+        set_n_features_in: bool = False,
+    ) -> _DecoratorType:
+
+        def wrapper(func):
```

Comment:
Is there a more descriptive name we can give to the wrapper? Asking because I spent a few minutes trying to understand if we were inspecting the signature on every function call or just at import time, and a different name might make it more obvious which we should expect. If there's no better name, we might just add a brief comment explaining what this is expected to do. It's already loads clearer than it was before, though!

Reply:
@wphicks Please have a look at whether the new naming scheme and expanded comments make things clearer.